advanced-data-types.md

```diff
@@ -0,0 +1,195 @@
+# Advanced data modeling with Realm .NET
+
+Realm's intuitive data model approach means that in most cases you don't even think of Realm models as entities - you just declare your POCOs, have them inherit from `RealmObject` and you're done. Now you have persistable models, with `INotifyPropertyChanged` capabilities all wired up, that are also "live" - i.e. every time you access a property, you get the latest state and not some snapshot from who knows how long ago. This is great and most of our users absolutely love the simplicity. Still, there are some use cases where being aware that you're working with a database can really bring your data models to the next level. In this blog post we'll evaluate 3 techniques you can apply to make your models fit your needs even better.
+
+## Constructor validation
+
+One of the core requirements of Realm is that all models need to have a parameterless constructor. This is needed because Realm needs to be able to instantiate an object without figuring out what the arguments to pass to the constructor. What not many people know is that you can make this parameterless constructor private to communicate expectations  to your callers. This means that if you have an `Person` class where you absolutely expect that a `Name` is provided upon object creation, you can have a public constructor with a `name` argument and a private parameterless one for use by Realm:
+
+```csharp
+class Person : RealmObject
+{
+    public string Name { get; set; }
+
+    public Person(string name)
+    {
+        ValidateName(name);
+
+        Name = name;
+    }
+
+    // This is used by Realm, even though it's private
+    private Person()
+    {
+    }
+}
+```
+
+And I know what some of you may be thinking - "oh no 😱 does that mean Realm uses the suuuuper slow reflection to create object instances". Fortunately, the answer is "no". Instead, at compile-time, Realm injects a nested helper class in each model that has a `CreateInstance` method. Since the helper class is nested in the model classes, it has access to private members and is thus able to invoke the private constructor.
+
+## Property access modifiers
+
+Similar to the point above - another relatively unknown feature of Realm is that persisted properties don't need to be public. You can either have the entire property be private or just one of the accessors. This synergizes nicely with the private constructor technique that we mentioned above. If you expose a constructor that explicitly validates the person's name, it would be fairly annoying to do all that work and have some code accidentally set the property to `null` the very next line. So it would make sense to make the setter of the name property above private:
+
```

Comment:
Is it worth mentioning here that they could also implement a custom setter validating the name on every change?

Reply:
Yes, it's below.